cachito/workers/pkg_managers/gomod.py

```diff
@@ -304,52 +281,149 @@ def resolve_gomod(app_source_path, request, dep_replacements=None, git_dir_path=
             )
             _merge_bundle_dirs(tmp_download_cache_dir, str(bundle_dir.gomod_download_dir))
 
-        log.info("Retrieving the list of packages")
-        package_list = run_gomod_cmd([*go_list, "-find", "./..."], run_params).splitlines()
+        go_list = ["go", "list", "-e"]
+        if not should_vendor:
+            # Make Go ignore the vendor dir even if there is one
+            go_list.extend(["-mod", "readonly"])
 
-        log.info("Retrieving the list of package level dependencies")
-        package_info = _load_list_deps(
-            run_gomod_cmd([*go_list, "-deps", "-json", "./..."], run_params)
+        main_module_name = run_gomod_cmd([*go_list, "-m"], run_params).rstrip()
+        main_module_version = get_golang_version(
+            main_module_name,
+            git_dir_path,
+            request["ref"],
+            update_tags=True,
+            subpath=(
+                None
+                if app_source_path == str(git_dir_path)
+                else app_source_path.replace(f"{git_dir_path}/", "")
+            ),
         )
+        main_module = {"type": "gomod", "name": main_module_name, "version": main_module_version}
+
+        def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[GoPackage]:
+            """Run go list -deps -json and return the parsed list of packages.
 
-        packages = []
-        processed_pkg_deps = set()
-        for pkg_name in package_list:
-            if pkg_name in processed_pkg_deps:
-                # Go searches for packages in directories through a top-down approach. If a toplevel
-                # package is already listed as a dependency, we do not list it here, since its
-                # dependencies would also be listed in the parent package
+            The "./..." pattern returns the list of packages compiled into the final binary.
+
+            The "all" pattern includes dependencies needed only for tests. Use it to get a more
+            complete module list (roughly matching the list of downloaded modules).
+            """
+            cmd = [*go_list, "-deps", "-json=ImportPath,Module,Standard,Deps", pattern]
+            return map(GoPackage.parse_obj, load_json_stream(run_gomod_cmd(cmd, run_params)))
+
+        package_modules = (
+            mod for pkg in go_list_deps("all") if (mod := pkg.module) and not mod.main
+        )
+        main_module_deps = _deduplicate_to_gomod_dicts(
+            chain(downloaded_modules, package_modules), deps_to_replace
+        )
```

Comment:
^ pretty much the only difference between this and [the cachi2 version](https://github.com/containerbuildsystem/cachi2/blob/8b7d3de429cef774151608beae0b93001e76d61f/cachi2/core/package_managers/gomod.py#L329-L332)

Reply:
The order of the arguments here is important.. if we switched to package_modules first, we'd miss the replacement info. Maybe add a comment to emphasize it.

Reply:
I am also fine with previous efficient one.

Reply:
Yeah, it still works. I think it's because what's reported in `go mod download -json`  already has the `.replace.path` in `.path`, so it will resolve correctly in `_get_name_and_version`. 

Nevermind, and sorry for the noise.